### ISSUE TITLE
Here are the required changes to support MacOS

### ISSUE THREAD

**Issue by phillipjohnston** [johnttaylor/pim#9]
MacOS can be supported with the following trivial modifications:

1. Alternate Semaphore implementation (see [the attached patch](https://github.com/johnttaylor/pim/files/9066062/apple_sem_support_patch.txt)). The implementation is derived from [my own Sempahore class](https://github.com/embvm/embvm-core/blob/master/src/os/posix/posix_semaphore.hpp), and the `pim/tests/Cpl/System/_0test/realtime/linux/gcc` tests passed when run.
1. Build system linking changes (I wasn't sure how to best make these, so I defer changes to the custom build system to you)
    1. Link against `-lc++` instead of `-lstdc++` (well, this is probably clang-specific not MacOS specific)
    1. Don't emit `--start-group` and `--end-group` linker arguments (also, this is clang-specific), e.g.,:
    ```diff
    -        startgroup = self._linker_libgroup_start if libs != '' else ''
    -        endgroup   = self._linker_libgroup_end   if libs != '' else ''
    +        startgroup = ''
    +        endgroup   = ''
    ```
1. Addressing #8, #7, #4.

Runs:

```
--- Your friendly neighborhood TShell. ---


$$$ threads

Name              R  ID          Native Hdl
----              -  --          ----------
PosixMain         Y  0x10b629dc0  0x10b629dc0
TShell            Y  0x70000bbb2000  0x70000bbb2000
NVRAM             Y  0x70000bb2f000  0x70000bb2f000
Algorithm         Y  0x70000bc35000  0x70000bc35000
HouseSim          Y  0x70000baac000  0x70000baac000

Total number of threads: 5
$$$ dm ls
activeIdt
activeSetpoint
airFilterAlert
airFilterOperationTime
comfortConfig
cycleInfo
deltaIdtError
deltaSetpoint
enabledSecondaryIdt
equipmentBeginTimes
equipmentConfig
fanMode
freezePiRefCnt
houseSimEnabled
idtAlarms
inhibitfRefCnt
loggingFileName
loopCounter
maxAirFilterHours
noActiveConditioningAlarm
operatingModeChanged
outdoorTemp
primaryRawIdt
pvInhibited
pvOut
relayOutputs
resetPiPulse
secondaryRawIdt
setpointChanged
setpoints
sumError
systemConfig
systemForcedOffRefCnt
systemOn
userCfgModeAlarm
userMode
vOutputs
whiteBox
$$$
```

[apple_sem_support_patch.txt](https://github.com/johnttaylor/pim/files/9066062/apple_sem_support_patch.txt)

**Comment by johnttaylor** [comment]
### Item:1
For the OS X/Apple specific issue for semaphores - the PIM approach is to provide a new platform specific (i.e OS X/Apple) of the target interfaces defined in the `src/Cpl/System` namespace.  In this case because there appears to be very little differences between the currently supplied Posix implementation and what is needed for OS X/Apple - it would be simply cloning the `src/Cpl/System/Posix/` directory to something like `src/Cpl/System/Apple/` (not sure what the best name would be) and then modify the `.../Apple/Semaphore.cpp` to contain the Apple specific implementation.  If you want to make this update, I will be happy to incorporate into the repo (especially since I am pretty sure I will not have a MacBook any time soon).

Note: I did some quick research on the change to use the GCD semaphores - and my concern is that the GCD semaphores are not safe to use in signal handlers, i.e. does not meet the defined semantics for `int su_signal( void ) noexcept;`.  That said, the Cpl C++ library is not really intended to be general purpose desktop C++ library so it is not the end-of-world if the su_signal() for OS X doesn't do anything.  Anyway - if you are interesting in providing the update - we chat about what to do about the su_signal() implementation.

### Item 2
I have addressed this in detail in issue: #4. 

### Item 3
See each of the issues from my responses.

**Comment by phillipjohnston** [comment]
> In this case because there appears to be very little differences between the currently supplied Posix implementation and what is needed for OS X/Apple - it would be simply cloning the src/Cpl/System/Posix/ directory to something like src/Cpl/System/Apple/ (not sure what the best name would be) and then modify the .../Apple/Semaphore.cpp to contain the Apple specific implementation.

Well, the ONLY differences are the one type - it is otherwise a POSIX system. That is why in my own solution I did not duplicate everything to make its own distinct platform. Then I would have to deal with manually maintaining any fixes/changes across two different locations.

As an alternate hopefully-PIM-friendly approach: what about compile-time file selection based on the OS you are building the simulator for? linux simulator builds one semaphore file, MacOS the other.

Overall though, you are clear that your project supports Linux and Windows! My intent was not to advocate for MacOS support, but to at least highlight how minor the changes *if* you wanted to support it After all, you have to maintain it over the long run! Maybe that's not ideal if you don't have access to a Mac.

**Comment by phillipjohnston** [comment]
> Agreed, but it's slippery slope. Today it is just one set of #ifdefs, then after six months there are dozens of #ifdefs.

Sure, some slopes are slippery. But it's been at least 5 years in my own implementation without any other variations required, so I am confident in my claim that it is the only difference with respect to both of our required OS abstractions.

> Anyway - if you are interesting in providing the update - we chat about what to do about the su_signal() implementation.

Did you have any notes on this that I should keep in mind when making the changes you described?

**Comment by phillipjohnston** [comment]
tldr: I will move my own implementation and your implementation over to mach/semaphore instead of GCD semaphore.

## Longer Version

I did a bit of poking around re: signal handler use due to the implications for my own simulator/framework. I did indeed miss that GCD semaphores cannot be used in signal handlers. I also looked into POSIX named semaphores on MacOS as an alternative, but like GCD sems they are also build on top of [mach/semaphores](https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/KernelProgramming/synchronization/synchronization.html) as well. Does that mean POSIX semaphores are unsafe to use in signal handlers? That would surprise me if it was a) the case, and b) not actually addressed after all these years.

I found [this discussion](https://github.com/adrienverge/openfortivpn/issues/105) which reaches similar conclusions and some additional insight. However, I was somewhat surprised at the conclusion to move to mach semaphores, because that actually still leaves me with many questions:

- Is the documentation for mach semaphores wrong, and they are actually safe to use in signal handlers? (seems likely)
- Does it mean that the POSIX implementation on MacOS for named semaphores are unsafe, fundamentally? (seems unlikely)
- Does that mean that GCD semaphores, also based on mac semaphores, are practically safe even though documented otherwise? (hard to say...)

Looking into the source code, it seems that the Mac POSIX-named/mach semaphore calls are entirely kernel side. GCD look like they're built in userspace on top of an atomic counter. So my conclusion is that the mach semaphores would be as safe as you can get on MacOS, but GCD would not be safe to use in a signal handler.

**Comment by johnttaylor** [comment]
I agree with the mach semaphores.  My quick research seem to indicate that mach semaphores work in signal handlers.
